src/main/java/org/apache/xml/security/keys/content/AgreementMethodImpl.java

```diff
@@ -0,0 +1,324 @@
+     * must be set later.
+     *
+     * @param algorithm the algorithm URI for the key agreement algorithm
+     */
+    public AgreementMethodImpl(Document doc, String algorithm) {
+        super(doc);
+
+        agreementMethodInformation = new LinkedList<>();
+        URI tmpAlgorithm;
+        try {
+            tmpAlgorithm = new URI(algorithm);
+        } catch (URISyntaxException ex) {
+            throw new IllegalArgumentException("Algorithm [" + algorithm + "] is not URI ", ex);
+        }
+        algorithmURI = tmpAlgorithm.toString();
+
+        setLocalAttribute(Constants._ATT_ALGORITHM, algorithmURI);
+    }
+
+    /**
+     * Constructor AgreementMethodImpl based on XML {@link Element}.
+     *
+     * @param element the XML {@link Element} containing AgreementMethod information
+     * @throws XMLSecurityException
+     */
+    public AgreementMethodImpl(Element element) throws XMLSecurityException {
+        super(element, EncryptionConstants.EncryptionSpecNS);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public byte[] getKANonce() {
+        return kaNonce;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setKANonce(byte[] kanonce) {
+        kaNonce = kanonce;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Iterator<Element> getAgreementMethodInformation() {
+        return agreementMethodInformation.iterator();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void addAgreementMethodInformation(Element info) {
+        agreementMethodInformation.add(info);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void removeAgreementMethodInformation(Element info) {
+        agreementMethodInformation.remove(info);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public KeyDerivationMethod getKeyDerivationMethod() throws XMLSecurityException {
+
+        if (keyDerivationMethod != null) {
+            return keyDerivationMethod;
+        }
+
+        Element keyDerivationMethodElement =
+                XMLUtils.selectXenc11Node(getElement().getFirstChild(), EncryptionConstants._TAG_KEYDERIVATIONMETHOD, 0);
+
+        if (keyDerivationMethodElement == null) {
+            return null;
+        }
+        keyDerivationMethod = new KeyDerivationMethodImpl(keyDerivationMethodElement, baseURI);
+
+
+        return keyDerivationMethod;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setKeyDerivationMethod(KeyDerivationMethod keyDerivationMethod) {
+        this.keyDerivationMethod = keyDerivationMethod;
+        if (keyDerivationMethod instanceof ElementProxy) {
+            appendSelf((ElementProxy) keyDerivationMethod);
+            addReturnToSelf();
+        } else {
+            LOG.log(Level.WARNING, "KeyDerivationMethod is set but is not an instance of ElementProxy. The DOM node is lost upon serialization.");
```

Comment:
Maybe the effective class of `keyDerivationMethod` could be part of the log message to help spot errors more easily?